File: cmd/admin/purge.py
import asyncio
import datetime
import discord
from discord.ext import commands
from utils.permissions.base_admin import BaseAdminCog
from utils.core.embeds import make_embed
from utils.core.emojis import EMOJIS
from utils.logging.mod_log import send_mod_log

GuildChannel = discord.TextChannel | discord.Thread | discord.VoiceChannel | discord.StageChannel


class Purge(BaseAdminCog):

    MAX_PURGE = 1000
    MAX_SCAN = 5000

    def __init__(self, bot: commands.Bot):
        self.bot = bot

    async def has_purge_permission(self, ctx: commands.Context) -> bool:
        guild = ctx.guild
        if guild is None:
            return False

        author = ctx.author
        if not isinstance(author, discord.Member):
            return False

        if await self._has_access(member=author, guild=guild, ctx=ctx):
            return True

        return author.guild_permissions.manage_messages

    async def _reply(self,
                     ctx: commands.Context,
                     title: str,
                     description: str,
                     level: str = "ERROR",
                     delete_after: int = 8):
        try:
            return await ctx.reply(embed=make_embed(title=title,
                                                    description=description,
                                                    level=level,
                                                    use_emoji=True),
                                   mention_author=False,
                                   delete_after=delete_after)
        except discord.HTTPException:
            return None

    async def _cleanup(self, ctx: commands.Context):
        try:
            await ctx.message.delete()
        except (discord.Forbidden, discord.NotFound, discord.HTTPException):
            pass

    async def resolve_target(self,
                             ctx: commands.Context) -> discord.Member | None:

        # MENTION DETECTION
        if ctx.message.mentions:
            member = ctx.message.mentions[0]
            if isinstance(member, discord.Member):
                return member
        # REPLY TARGET DETECTION
        reference = ctx.message.reference
        if (reference and isinstance(reference.resolved, discord.Message)):
            author = reference.resolved.author
            if isinstance(author, discord.Member):
                return author
        return None

    async def collect_messages(
            self,
            *,
            channel: GuildChannel,
            amount: int,
            member: discord.Member | None,
            exclude: int | None = None) -> list[discord.Message]:
        messages: list[discord.Message] = []
        if member:
            scan_limit = self.MAX_SCAN
        else:
            scan_limit = min(amount + 50, self.MAX_SCAN)
        async for msg in channel.history(limit=scan_limit):
            if exclude and msg.id == exclude:
                continue
            if msg.pinned:
                continue
            if member and msg.author.id != member.id:
                continue
            messages.append(msg)
            if len(messages) >= amount:
                break

        return messages

    async def delete_messages(self, *, channel: GuildChannel,
                              messages: list[discord.Message]) -> int:
        now = discord.utils.utcnow()
        fourteen_days = datetime.timedelta(days=14)
        young: list[discord.Message] = []
        old: list[discord.Message] = []

        for msg in messages:
            if now - msg.created_at < fourteen_days:
                young.append(msg)
            else:
                old.append(msg)

        deleted = 0

        # Bulk delete for fresh text objects
        while young:
            batch = young[:100]
            young = young[100:]

            try:
                await channel.delete_messages(batch)
                deleted += len(batch)
            except discord.HTTPException:
                pass

        for msg in old:
            try:
                await msg.delete()
                deleted += 1
                await asyncio.sleep(0.35)
            except discord.HTTPException:
                pass

        return deleted

    @commands.command(name="purge")
    @commands.guild_only()
    @commands.cooldown(1, 3, commands.BucketType.member)
    @commands.max_concurrency(1, per=commands.BucketType.guild, wait=False)
    async def purge(self, ctx: commands.Context, *args):
        await self._cleanup(ctx)

        guild = ctx.guild
        channel = ctx.channel

        if guild is None:
            return

        if not isinstance(channel,
                          (discord.TextChannel, discord.Thread,
                           discord.VoiceChannel, discord.StageChannel)):
            return

        moderator = ctx.author
        if not isinstance(moderator, discord.Member):
            return

        if not await self.has_purge_permission(ctx):
            return await self._reply(
                ctx,
                "Permission Denied",
                "You do not have permission to use this command.",
                level="ERROR",
            )

        bot_member = guild.me
        if bot_member is None:
            return

        channel_perms = channel.permissions_for(bot_member)
        if not channel_perms.manage_messages:
            return await self._reply(
                ctx,
                "Missing Permissions",
                "I need the `Manage Messages` permission in this target channel.",
                level="ERROR",
            )

        member = None
        amount = None

        if not args:
            return await self._reply(
                ctx,
                "Invalid Usage",
                ("`purge <amount>`\n"
                 "`purge @user <amount>`\n"
                 "`reply + purge <amount>`"),
                level="WARNING",
            )

        member = await self.resolve_target(ctx)

        for arg in args:
            if arg.isdigit():
                amount = int(arg)
                break

        if amount is None:
            return await self._reply(
                ctx,
                "Missing Amount",
                "Please provide a valid number of messages to purge.",
                level="WARNING",
            )

        if amount < 1 or amount > self.MAX_PURGE:
            return await self._reply(
                ctx,
                "Invalid Amount",
                f"Amount must be between 1 and {self.MAX_PURGE}.",
                level="WARNING",
            )

        messages_to_delete = await self.collect_messages(
            channel=channel,
            amount=amount,
            member=member,
            exclude=ctx.message.id,
        )

        if not messages_to_delete:
            return await self._reply(
                ctx,
                "Purge Complete",
                "No messages matched your criteria to delete.",
                level="INFO",
                delete_after=5,
            )

        deleted_count = await self.delete_messages(
            channel=channel,
            messages=messages_to_delete,
        )

        # USER NOTIFICATION
        target_string = f"from {member.mention}" if member else "from this channel"
        await self._reply(
            ctx,
            "Messages Purged",
            f"Successfully deleted **{deleted_count}** messages {target_string}.",
            level="SUCCESS",
            delete_after=5,
        )

        await send_mod_log(
            guild=guild,
            category="CONFIG",
            title="Channel Clean Purge",
            description=
            f"Bulk deleted **{deleted_count}** messages in {channel.mention}.",
            level="SUCCESS",
            actor=moderator,
            target=member,
            extra_fields={
                "Channel ID": channel.id,
                "Requested Target": amount,
                "Actual Deleted": deleted_count
            })

    @purge.error
    async def purge_error(self, ctx: commands.Context,
                          error: commands.CommandError):
        if isinstance(error, commands.CommandOnCooldown):
            await self._cleanup(ctx)
            return await self._reply(
                ctx,
                "Command on Cooldown",
                f"Please allow the database to settle. Retry in **{error.retry_after:.1f}s**.",
                level="WARNING",
                delete_after=4)

        # Active operation race conditions protection
        if isinstance(error, commands.MaxConcurrencyReached):
            await self._cleanup(ctx)
            return await self._reply(
                ctx,
                "System Busy",
                "An administrative purge operation is currently running here. Please standby.",
                level="WARNING",
                delete_after=5)
        raise error


async def setup(bot: commands.Bot):
    await bot.add_cog(Purge(bot))

File: src/commands/admin/purge.js
import { PermissionFlagsBits, SlashCommandBuilder } from "discord.js";
import { createCommand } from "../../core/command.js";
import { makeEmbed } from "../../core/embeds.js";
import { EMOJIS } from "../../core/emojis.js";
import { sendModLog } from "../../utils/modLog.js";

const MAX_PURGE = 1000;
const MAX_SCAN = 5000;

// Helper to standardise short-lived replies matching python version
async function _reply(ctx, title, description, level = "ERROR", deleteAfterSecs = 8) {
  const emojiKey = level.toLowerCase();
  const systemEmoji = EMOJIS.get(emojiKey) || EMOJIS.get("warning") || "";
  const formattedDescription = systemEmoji ? `${systemEmoji} ${description}`.trim() : description;

  const embed = makeEmbed({
    title,
    description: formattedDescription,
    level
  });

  const replyOptions = { embeds: [embed] };
  if (ctx.isInteraction) {
    replyOptions.ephemeral = true;
  }

  const msg = await ctx.reply(replyOptions);

  // Auto-delete message command replies after N seconds
  if (!ctx.isInteraction && msg && deleteAfterSecs > 0) {
    setTimeout(() => {
      msg.delete().catch(() => null);
    }, deleteAfterSecs * 1000);
  }
  return msg;
}

const slashBuilder = new SlashCommandBuilder()
  .setName("purge")
  .setDescription("Bulk delete up to 1000 messages from the current channel")
  .addIntegerOption((opt) =>
    opt
      .setName("amount")
      .setDescription(`Number of messages to delete (1-${MAX_PURGE})`)
      .setRequired(true)
      .setMinValue(1)
      .setMaxValue(MAX_PURGE)
  )
  .addUserOption((opt) => opt.setName("user").setDescription("Filter messages by specific user").setRequired(false));

export default createCommand({
  name: "purge",
  description: "Cog for performing bulk message deletion in guild channels.",
  category: "Admin",
  aliases: ["clear"],
  modOnly: true,
  requiredPermission: PermissionFlagsBits.ManageMessages,
  slashBuilder,

  async execute(ctx) {
    const { channel, guild, message, client, isInteraction } = ctx;
    if (!channel || !channel.messages) {
      return await ctx.reply({ content: "Cannot purge messages in this channel type.", ephemeral: true });
    }

    // Bot Permission Check
    if (!channel.permissionsFor(guild.members.me).has(PermissionFlagsBits.ManageMessages)) {
      return await _reply(ctx, "Missing Permissions", "I need the `Manage Messages` permission in this target channel.", "ERROR");
    }

    let targetUser = null;
    let amount = undefined;

    if (isInteraction) {
      amount = parseInt(ctx.options.amount);
      if (ctx.options.user) {
        const targetId = typeof ctx.options.user === "string" ? ctx.options.user : ctx.options.user.id;
        targetUser = await client.users.fetch(targetId).catch(() => null);
      }
    } else {
      // 1. Resolve Target User from mentions or reply reference
      if (message.mentions.users.size > 0) {
        targetUser = message.mentions.users.first();
      } else if (message.reference && message.reference.messageId) {
        try {
          const refMsg = await channel.messages.fetch(message.reference.messageId);
          if (refMsg) targetUser = refMsg.author;
        } catch (e) {
          // Ignore fetch error
        }
      }

      // 2. Resolve Amount from arguments
      if (ctx.args) {
        for (const arg of ctx.args) {
          if (/^\d+$/.test(arg)) {
            amount = parseInt(arg, 10);
            break;
          }
        }
      }

      // 3. Delete invocation message
      await message.delete().catch(() => null);
    }

    if (!amount || isNaN(amount)) {
      const warningIcon = EMOJIS.get("warning") || "⚠️";
      return await _reply(
        ctx,
        "Invalid Usage",
        `${warningIcon} Correct syntax:\n• \`purge <amount>\`\n• \`purge @user <amount>\`\n• \`reply + purge <amount>\``,
        "WARNING"
      );
    }

    if (amount < 1 || amount > MAX_PURGE) {
      return await _reply(ctx, "Invalid Amount", `Amount must be between 1 and ${MAX_PURGE}.`, "WARNING");
    }

    // Acknowledge interaction quickly to prevent timeout during long scans
    if (isInteraction) {
      await ctx.defer({ ephemeral: true });
    }

    // ---------------------------------------------------------
    // Collect Messages
    // ---------------------------------------------------------
    const messagesToDelete = [];
    const scanLimit = targetUser ? MAX_SCAN : Math.min(amount + 50, MAX_SCAN);
    
    let lastId = undefined;
    let fetched = 0;

    while (messagesToDelete.length < amount && fetched < scanLimit) {
      const fetchAmount = Math.min(100, scanLimit - fetched);
      if (fetchAmount <= 0) break;

      const batch = await channel.messages.fetch({ limit: fetchAmount, before: lastId }).catch(() => null);
      if (!batch || batch.size === 0) break;
      
      fetched += batch.size;
      lastId = batch.last().id;

      for (const msg of batch.values()) {
        if (messagesToDelete.length >= amount) break;
        
        // Exclude pinned messages
        if (msg.pinned) continue;
        // Exclude the command message itself (if it somehow wasn't deleted)
        if (!isInteraction && message && msg.id === message.id) continue;
        // Exclude if filtered by user
        if (targetUser && msg.author.id !== targetUser.id) continue;
        
        messagesToDelete.push(msg);
      }
    }

    if (messagesToDelete.length === 0) {
      return await _reply(ctx, "Purge Complete", "No messages matched your criteria to delete.", "INFO", 5);
    }

    // ---------------------------------------------------------
    // Delete Messages (Handling 14-day limit)
    // ---------------------------------------------------------
    const now = Date.now();
    const fourteenDays = 14 * 24 * 60 * 60 * 1000;
    const young = [];
    const old = [];

    for (const msg of messagesToDelete) {
      // Bulk delete only supports messages younger than 14 days
      if (now - msg.createdTimestamp < fourteenDays) {
        young.push(msg);
      } else {
        old.push(msg);
      }
    }

    let deletedCount = 0;

    // Delete young messages in bulk (max 100 per API call)
    for (let i = 0; i < young.length; i += 100) {
      const chunk = young.slice(i, i + 100);
      try {
        const deleted = await channel.bulkDelete(chunk, true);
        deletedCount += deleted.size;
      } catch (e) {
        // Safe fail
      }
    }

    // Delete old messages one by one with rate-limit respecting delay
    for (const msg of old) {
      try {
        await msg.delete();
        deletedCount++;
        // 350ms delay matching python equivalent
        await new Promise(r => setTimeout(r, 350));
      } catch (e) {
        // Safe fail
      }
    }

    // ---------------------------------------------------------
    // Response & Mod Log
    // ---------------------------------------------------------
    const targetString = targetUser ? `from <@${targetUser.id}>` : "from this channel";
    const successIcon = EMOJIS.get("success") || "✅";
    
    await _reply(ctx, "Messages Purged", `${successIcon} Successfully deleted **${deletedCount}** messages ${targetString}.`, "SUCCESS", 5);

    try {
      await sendModLog({
        guild,
        category: "CONFIG", // Matching Python format
        title: "Channel Clean Purge",
        description: `Bulk deleted **${deletedCount}** messages in <#${channel.id}>.`,
        level: "SUCCESS",
        actor: ctx.user,
        extraFields: {
          "Channel ID": channel.id,
          "Requested Target": String(amount),
          "Actual Deleted": String(deletedCount),
        }
      });
    } catch (e) {
      // Ignore mod log failures
    }
  }
});
